experimental/packages/opentelemetry-sdk-metrics-base/src/state/AsyncMetricStorage.ts

```diff
@@ -0,0 +1,100 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { HrTime } from '@opentelemetry/api';
+import { ObservableCallback } from '@opentelemetry/api-metrics-wip';
+import { Accumulation, Aggregator } from '../aggregator/types';
+import { View } from '../view/View';
+import { createInstrumentDescriptorWithView, InstrumentDescriptor } from '../InstrumentDescriptor';
+import { AttributesProcessor } from '../view/AttributesProcessor';
+import { MetricStorage } from './MetricStorage';
+import { InstrumentationLibrary } from '@opentelemetry/core';
+import { Resource } from '@opentelemetry/resources';
+import { MetricData } from '../export/MetricData';
+import { DeltaMetricProcessor } from './DeltaMetricProcessor';
+import { TemporalMetricProcessor } from './TemporalMetricProcessor';
+import { Maybe } from '../utils';
+import { MetricCollectorHandle } from './MetricCollector';
+import { ObservableResult } from '../ObservableResult';
+import { AttributeHashMap } from './HashMap';
+
+/**
+ * Internal interface.
+ *
+ * Stores and aggregates {@link MetricData} for asynchronous instruments.
+ */
+export class AsyncMetricStorage<T extends Maybe<Accumulation>> implements MetricStorage {
+  private _deltaMetricStorage: DeltaMetricProcessor<T>;
+  private _temporalMetricStorage: TemporalMetricProcessor<T>;
+
+  constructor(
+    private _instrumentDescriptor: InstrumentDescriptor,
+    aggregator: Aggregator<T>,
+    private _attributesProcessor: AttributesProcessor,
+    private _callback: ObservableCallback
+  ) {
+    this._deltaMetricStorage = new DeltaMetricProcessor(aggregator);
+    this._temporalMetricStorage = new TemporalMetricProcessor(aggregator);
+  }
+
+  private _record(measurements: AttributeHashMap<number>) {
+    const processed = new AttributeHashMap<number>();
+    Array.from(measurements.entries()).forEach(([attributes, value]) => {
+      processed.set(this._attributesProcessor.process(attributes), value);
+    });
+    this._deltaMetricStorage.batchCumulate(processed);
+  }
+
+  /**
+   * Collects the metrics from this storage. The ObservableCallback is invoked
+   * during the collection.
+   *
+   * Note: This is a stateful operation and may reset any interval-related
+   * state for the MetricCollector.
+   */
+  async collect(
+    collector: MetricCollectorHandle,
+    collectors: MetricCollectorHandle[],
+    resource: Resource,
+    instrumentationLibrary: InstrumentationLibrary,
+    sdkStartTime: HrTime,
+    collectionTime: HrTime,
+  ): Promise<Maybe<MetricData>> {
+    const observableResult = new ObservableResult();
+    // TODO: timeout with callback
+    await this._callback(observableResult);
```

Comment:
I'd believe so. Since `MetricReader.collect` already has a timeout option, I'd believe the timeout should be applied on async metric observations, as the timeout is not reasonable for synchronous instrumentation in JavaScript (we can't cancel or interrupt sync operations with pure JavaScript) and other parts of a collection (they are all synchronous).

Reply:
Sounds good!